src/go/rpk/pkg/cli/topic/list_test.go

```diff
@@ -191,3 +200,180 @@ func TestEmptyTopicList(t *testing.T) {
+	cmd.SetArgs([]string{"list"})
+	output = captureOutput(func() {
+		cmd.Execute()
+	})
+
+	require.Contains(t, output, "topic1")
+	require.Contains(t, output, "topic2")
+	require.Contains(t, output, "topic3")
+	require.Contains(t, output, "NAME")
+	require.Contains(t, output, "PARTITIONS")
+	require.Contains(t, output, "REPLICAS")
+}
+
+func TestTopicListCommand_RegexFiltering(t *testing.T) {
+	tests := []struct {
+		name           string
+		topics         []string
+		regexPattern   string
+		expectedTopics []string
+		notExpected    []string
+	}{
+		{
+			name:           "match single pattern",
+			topics:         []string{"user-events", "user-login", "system-metrics", "system-logs"},
+			regexPattern:   "user-.*",
+			expectedTopics: []string{"user-events", "user-login"},
+			notExpected:    []string{"system-metrics", "system-logs"},
+		},
+		{
+			name:           "match all with wildcard",
+			topics:         []string{"topic1", "topic2"},
+			regexPattern:   ".*",
+			expectedTopics: []string{"topic1", "topic2"},
+			notExpected:    []string{},
+		},
+		{
+			name:           "exact match",
+			topics:         []string{"exact-topic", "exact-topic-2", "other"},
+			regexPattern:   "exact-topic",
+			expectedTopics: []string{"exact-topic"},
+			notExpected:    []string{"exact-topic-2", "other"},
+		},
+		{
+			name:           "no match returns empty",
+			topics:         []string{"topic1", "topic2", "topic3"},
+			regexPattern:   "nonexistent-.*",
+			expectedTopics: []string{},
+			notExpected:    []string{"topic1", "topic2", "topic3"},
+		},
+		{
+			name:           "partial match fails due to incorrect regex",
+			topics:         []string{"user-events", "events-user"},
+			regexPattern:   "events",
+			expectedTopics: []string{},
+			notExpected:    []string{"user-events", "events-user"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cmd, _ := setupTestCommand(t, tt.topics...)
+
+			// Execute list command with regex
+			cmd.SetArgs([]string{"list", "--regex", tt.regexPattern})
+			output := captureOutput(func() {
+				cmd.Execute()
+			})
+
+			// Check expected topics are in output
+			for _, topic := range tt.expectedTopics {
+				require.Contains(t, output, topic,
+					"Expected topic %s to be in output for regex %s", topic, tt.regexPattern)
+			}
+
+			// Check unexpected topics are NOT in output
+			for _, topic := range tt.notExpected {
+				require.NotContains(t, output, topic,
+					"Topic %s should NOT be in output for regex %s", topic, tt.regexPattern)
+			}
+
+			// Special check for the bug fix case
+			if tt.name == "no match returns empty" {
+				// Should only contain headers, no actual topics
+				lines := strings.Split(strings.TrimSpace(output), "\n")
+				require.Equal(t, 1, len(lines),
+					"Non-matching regex should only show header line, not all topics")
+				require.Contains(t, lines[0], "NAME")
+				require.Contains(t, lines[0], "PARTITIONS")
+			}
```

Comment:
Alernatively we could add a function to the test struct and define it in the table. This way if we add another case we won't have to match per name over here.